java-squid/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -125,4 +138,80 @@ public String toString() {
     return "{" + values.toString() + "}  {" + constraints.toString() + "}" + " { " + stack.toString() + " }";
   }
 
+  public ProgramState addConstraint(SymbolicValue symbolicValue, Object constraint) {
+    PMap<SymbolicValue, Object> newConstraints = constraints.put(symbolicValue, constraint);
+    if (newConstraints != constraints) {
+      return new ProgramState(this, newConstraints);
+    }
+    return this;
+  }
+
+  ProgramState put(Symbol symbol, SymbolicValue value) {
+    if (symbol.isUnknown()) {
+      return this;
+    }
+    PMap<Symbol, SymbolicValue> newValues = values.put(symbol, value);
+    if (newValues != values) {
+      return new ProgramState(newValues, constraints, visitedPoints, stack);
+    }
+    return this;
+  }
+
+  public ProgramState resetFieldValues(ConstraintManager constraintManager) {
+    final List<VariableTree> variableTrees = new ArrayList<>();
+    values.forEach(new PMap.Consumer<Symbol, SymbolicValue>() {
+      @Override
+      public void accept(Symbol symbol, SymbolicValue value) {
+        if (isField(symbol)) {
+          VariableTree variable = ((Symbol.VariableSymbol) symbol).declaration();
+          if (variable != null) {
+            variableTrees.add(variable);
+          }
+        }
+      }
+    });
+    if (variableTrees.isEmpty()) {
+      return this;
+    }
+    PMap<Symbol, SymbolicValue> newValues = values;
+    for (VariableTree variableTree : variableTrees) {
+      newValues = newValues.put(variableTree.symbol(), constraintManager.createSymbolicValue(variableTree));
+    }
+    return new ProgramState(newValues, constraints, visitedPoints, stack);
+  }
+
+  private static boolean isField(Symbol symbol) {
+    return symbol.isVariableSymbol() && !symbol.owner().isMethodSymbol();
+  }
+
+  public boolean canReach(final SymbolicValue symbolicValue) {
+    final boolean[] reachable = new boolean[1];
+    values.forEach(new PMap.Consumer<Symbol, SymbolicValue>() {
+      @Override
+      public void accept(Symbol symbol, SymbolicValue value) {
+        if (value == symbolicValue) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)